README.md
## interval_time_selector

Interval time picker interval minutes, specific hour. This package help us to select time of day,
allow us to select specific hours selectable and minute interval this package is update on the
package [day_night_time_picker](https://pub.dev/packages/day_night_time_picker). We use this package
and modified the code to fit to select specific hours.

## Features

Help you to select specific hours selectable and minute interval.

## Screenshots

<a href="#screenshots">
<img src="https://github.com/MohamadAbdUlaziz938/interval-time-selector/blob/master/screenshots/1.png" width="200px">
</a>
<a href="#screenshots">
<img src="https://github.com/MohamadAbdUlaziz938/interval-time-selector/blob/master/screenshots/2.gif" width="200px">
</a>


## 💻 Installation

In the dependencies: section of your pubspec.yaml, add the following line:

```
interval_time_selector:<latest_version>
```

Import in your project

```
import 'package:interval_time_selector/show_time_picker/day_time_picker.dart';
```

## Usage

```
            Navigator.of(context).push(
                      showPicker(
                        workingHours: [0, 1, 9, 12, 11, 12, 13, 15, 19, 23],
                        context: context,
                        value: _time,
                        ascending: true,
                        onChange: onTimeChanged,
                        maxMinuteAtMaximumHour: 40,
                        minMinuteAtMinimumHour: 20,
                        onChangeDateTime: (DateTime dateTime) {
                          debugPrint("[debug datetime]:  $dateTime");
                        },
                      ),
           );
```

Longer examples find in `/example`
folder [example](https://github.com/MohamadAbdUlaziz938/interval-time-selector/tree/master/example).

## 👨 Updated By

Mohamad Abd-Ulaziz <a href="mailto:mohamad.samer.abdulaziz@gmail.com">Mail</a>

## Getting started

For help getting started with Flutter, view our
online [documentation](https://docs.flutter.dev/development/packages-and-plugins/developing-packages)
.




CHANGELOG.md
## 0.0.1

* Initial release specific hour, minute selector

## 0.0.2

* fix bugs, add gif



test/interval_time_picker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:interval_time_picker/interval_time_picker.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/show_time_picker/utils/utils.dart
import 'constant.dart';

double mapRange(
  double value,
  double iMin,
  double iMax, [
  double oMin = 0,
  double oMax = 1,
]) {
  return ((value - iMin) * (oMax - oMin)) / (iMax - iMin) + oMin;
}

int getIntFromMinuteIntervalEnum(MinuteInterval? interval) {
  switch (interval) {
    case MinuteInterval.FIVE:
      return 5;
    case MinuteInterval.TEN:
      return 10;
    case MinuteInterval.FIFTEEN:
      return 15;
    case MinuteInterval.THIRTY:
      return 30;
    default:
      return 5;
  }
}

/// Map MinuteInterval enum to division values
int getMinuteDivisions(int diff, MinuteInterval? interval) {
  switch (interval) {
    case MinuteInterval.FIVE:
      // 12
      return (diff / 5).round();
    case MinuteInterval.TEN:
      // 6
      return (diff / 10).round();
    case MinuteInterval.FIFTEEN:
      // 4
      return (diff / 15).round();
    case MinuteInterval.THIRTY:
      // 2
      return (diff / 30).round();
    default:
      return diff;
  }
}

double getMinMinute(double? minMinute) {
  return 0;
}

/// Get the minimum minute from interval
double getMinMinuteStepper(double? minMinute, MinuteInterval? interval) {
  if (minMinute == 0) {
    return 0;
  }
  int step = getIntFromMinuteIntervalEnum(interval);

  double min = -1;
  double i = 1;
  while (min < 0) {
    double val = i * step;
    if (val >= minMinute!) {
      min = val;
    }
    i++;
  }
  return min;
}

/// Get the maximum minute from interval
double getMaxMinute(double maxMinute, double minMinute) {
  double minPeriod = minMinute;
  double maxPeriod = maxMinute;
  final diff = (maxPeriod - minPeriod) / 5;
  return diff;
}
// double getMaxMinute(double? maxMinute, MinuteInterval? interval) {
//   if (maxMinute == 59) {
//     return 59;
//   }
//   int step = getIntFromMinuteIntervalEnum(interval);
//
//   double max = 60;
//   double i = 1;
//   while (max > maxMinute!) {
//     double val = 60 - (i * step);
//     if (val <= maxMinute) {
//       max = val;
//     }
//     i++;
//   }
//   return max;
// }

/// Generate a List of minutes
List<int?> generateMinutes(int divisions, MinuteInterval? interval, min, max) {
  final minutes = List<int?>.generate(divisions + 1, (index) {
    final val = min.round() + (getIntFromMinuteIntervalEnum(interval) * index);
    if (val >= max) {
      return max.round();
    }
    return val;
  });
  return minutes;
}

/// Generate a List of hours
List<int?> generateHours(int divisions, min, max) {
  final hours = List<int?>.generate(divisions, (index) {
    final val = min.round() + index;
    if (val >= max) {
      return max.round();
    }
    return val;
  });
  return hours;
}


lib/show_time_picker/day_time_picker.dart
import 'package:flutter/material.dart';
import 'package:interval_time_selector/show_time_picker/utils/utils.dart';

import 'state/state_container.dart';
import 'state/time.dart';
import 'utils/constant.dart';
import 'widgets/android_builder/day_time_picker_android.dart';

PageRouteBuilder showPicker({
  BuildContext? context,
  required TimeOfDay value,
  required List<int> workingHours,
  required void Function(TimeOfDay) onChange,
  void Function(DateTime)? onChangeDateTime,
  void Function()? onCancel,
  bool is24HrFormat = false,
  Color? accentColor,
  Color? unselectedColor,
  String cancelText = "Cancel",
  String okText = "Ok",
  Image? sunAsset,
  Image? moonAsset,
  bool blurredBackground = false,
  bool ltrMode = true,
  Color barrierColor = Colors.black45,
  double? borderRadius,
  double? elevation,
  EdgeInsets? dialogInsetPadding =
      const EdgeInsets.symmetric(horizontal: 40.0, vertical: 24.0),
  bool barrierDismissible = true,
  bool iosStylePicker = false,
  bool displayHeader = true,
  String hourLabel = 'hours',
  String minuteLabel = 'minutes',
  bool disableMinute = false,
  bool disableHour = false,
  double minMinuteAtMinimumHour = 0,
  double maxMinuteAtMaximumHour = 0,
  bool ascending = true,
  ThemeData? themeData,
  bool focusMinutePicker = false,
  TextStyle okStyle = const TextStyle(fontWeight: FontWeight.bold),
  TextStyle cancelStyle = const TextStyle(fontWeight: FontWeight.bold),
  ButtonStyle? buttonStyle,
  ButtonStyle? cancelButtonStyle,
  double? buttonsSpacing,
}) {
  assert((workingHours.isNotEmpty), "Working must be not empty");
  assert((minMinuteAtMinimumHour >= 0),
      "minMinuteAtCurrentHour must be equal or greater than zero");

  assert((maxMinuteAtMaximumHour >= 0),
      "maxMinuteAtMaximumHour must be equal or greater than zero");
  assert(!(disableHour == true && disableMinute == true),
      "Both \"disableMinute\" and \"disableHour\" cannot be true.");

  assert(!(disableMinute == true && focusMinutePicker == true),
      "Both \"disableMinute\" and \"focusMinutePicker\" cannot be true.");

  /// sort working hours - select max hour, min hour
  double minHour;
  double maxHour;
  if (ascending) {
    workingHours.sort();
    minHour = workingHours.first.toDouble();
    maxHour = workingHours.last.toDouble();
  } else {
    workingHours.sort(
      (a, b) => b.compareTo(a),
    );

    minHour = workingHours.first.toDouble();
    maxHour = workingHours.last.toDouble();
  }
  final selectedTimeValue = Time.fromTimeOfDay(value);

  return PageRouteBuilder(
    pageBuilder: (context, _, __) {
      {
        return Theme(
          data: themeData ?? Theme.of(context),
          child: const DayNightTimePickerAndroid(),
        );
      }
    },
    transitionDuration: const Duration(milliseconds: 200),
    transitionsBuilder: (context, anim, secondAnim, child) => SlideTransition(
      position: anim.drive(
        Tween(
          begin: const Offset(0, 0.15),
          end: const Offset(0, 0),
        ).chain(
          CurveTween(curve: Curves.ease),
        ),
      ),
      child: FadeTransition(
        opacity: anim,
        child: TimeModelBinding(
          workingHours: workingHours,
          maxMinuteAtMaximumHour: maxMinuteAtMaximumHour,
          minMinuteAtMinimumHour: minMinuteAtMinimumHour,
          selectedTime: selectedTimeValue,
          isInlineWidget: false,
          onChange: onChange,
          onChangeDateTime: onChangeDateTime,
          onCancel: onCancel,
          is24HrFormat: is24HrFormat,
          displayHeader: displayHeader,
          accentColor: accentColor,
          unselectedColor: unselectedColor,
          cancelText: cancelText,
          okText: okText,
          sunAsset: sunAsset,
          moonAsset: moonAsset,
          blurredBackground: blurredBackground,
          borderRadius: borderRadius,
          elevation: elevation,
          dialogInsetPadding: dialogInsetPadding,
          minuteInterval: MinuteInterval.FIVE,
          disableMinute: disableMinute,
          disableHour: disableHour,
          maxHour: maxHour,
          minHour: minHour,
          focusMinutePicker: focusMinutePicker,
          okStyle: okStyle,
          cancelStyle: cancelStyle,
          buttonStyle: buttonStyle,
          cancelButtonStyle: cancelButtonStyle,
          buttonsSpacing: buttonsSpacing,
          hourLabel: hourLabel,
          minuteLabel: minuteLabel,
          ltrMode: ltrMode,
          child: child,
        ),
      ),
    ),
    barrierDismissible: barrierDismissible,
    opaque: false,
    barrierColor: barrierColor,
  );
}

Widget createInlinePicker({
  BuildContext? context,
  void Function(DateTime)? onChangeDateTime,
  required TimeOfDay value,
  required void Function(TimeOfDay) onChange,
  void Function()? onCancel,
  bool is24HrFormat = false,
  Color? accentColor,
  Color? unselectedColor,
  String cancelText = "Cancel",
  String okText = "Ok",
  bool isOnChangeValueMode = false,
  bool ltrMode = true,
  Image? sunAsset,
  Image? moonAsset,
  bool blurredBackground = false,
  Color barrierColor = Colors.black45,
  double? borderRadius,
  double? elevation,
  EdgeInsets? dialogInsetPadding =
      const EdgeInsets.symmetric(horizontal: 40.0, vertical: 24.0),
  bool barrierDismissible = true,
  bool iosStylePicker = false,
  String hourLabel = 'hours',
  String minuteLabel = 'minutes',
  bool disableMinute = false,
  bool disableHour = false,
  bool displayHeader = true,
  ThemeData? themeData,
  bool focusMinutePicker = false,
  bool ascending = true,
  // Infinity is used so that we can assert whether or not the user actually set a value
  double minHour = double.infinity,
  double minMinuteAtMinimumHour = 0,
  double maxHour = double.infinity,
  required List<int> workingHours,
  TextStyle okStyle = const TextStyle(fontWeight: FontWeight.bold),
  TextStyle cancelStyle = const TextStyle(fontWeight: FontWeight.bold),
  ButtonStyle? buttonStyle,
  ButtonStyle? cancelButtonStyle,
  double maxMinuteAtMaximumHour = 0,
  double? buttonsSpacing,
}) {
  assert((workingHours.isNotEmpty), "Working must be not empty");
  assert((minMinuteAtMinimumHour >= 0),
      "minMinuteAtCurrentHour must be equal or greater than zero");

  assert((maxMinuteAtMaximumHour >= 0),
      "maxMinuteAtMaximumHour must be equal or greater than zero");
  assert(!(disableHour == true && disableMinute == true),
      "Both \"disableMinute\" and \"disableHour\" cannot be true.");

  assert(!(disableMinute == true && focusMinutePicker == true),
      "Both \"disableMinute\" and \"focusMinutePicker\" cannot be true.");

  /// sort working hours - select max hour, min hour
  double minHour;
  double maxHour;
  if (ascending) {
    workingHours.sort();
    minHour = workingHours.first.toDouble();
    maxHour = workingHours.last.toDouble();
  } else {
    workingHours.sort(
      (a, b) => b.compareTo(a),
    );

    minHour = workingHours.first.toDouble();
    maxHour = workingHours.last.toDouble();
  }
  final selectedTimeValue = Time.fromTimeOfDay(value);
  return TimeModelBinding(
    workingHours: ascending
        ? (workingHours..sort())
        : (workingHours
          ..sort(
            (a, b) => b.compareTo(a),
          )),
    maxMinuteAtMaximumHour: maxMinuteAtMaximumHour,
    minMinuteAtMinimumHour: minMinuteAtMinimumHour,
    onChange: onChange,
    selectedTime: selectedTimeValue,
    onChangeDateTime: onChangeDateTime,
    onCancel: onCancel,
    is24HrFormat: is24HrFormat,
    accentColor: accentColor,
    unselectedColor: unselectedColor,
    cancelText: cancelText,
    okText: okText,
    sunAsset: sunAsset,
    moonAsset: moonAsset,
    blurredBackground: blurredBackground,
    borderRadius: borderRadius,
    elevation: elevation,
    dialogInsetPadding: dialogInsetPadding,
    minuteInterval: MinuteInterval.FIVE,
    disableMinute: disableMinute,
    disableHour: disableHour,
    maxHour: maxHour,
    minHour: minHour,
    isInlineWidget: true,
    displayHeader: displayHeader,
    isOnValueChangeMode: isOnChangeValueMode,
    focusMinutePicker: focusMinutePicker,
    okStyle: okStyle,
    cancelStyle: cancelStyle,
    hourLabel: hourLabel,
    minuteLabel: minuteLabel,
    buttonStyle: buttonStyle,
    cancelButtonStyle: cancelButtonStyle,
    buttonsSpacing: buttonsSpacing,
    ltrMode: ltrMode,
    child: Builder(
      builder: (context) {
        {
          return Builder(
            builder: (context) {
              return Theme(
                data: themeData ?? Theme.of(context),
                child: const DayNightTimePickerAndroid(),
              );
            },
          );
        }
      },
    ),
  );
}


lib/show_time_picker/widgets/sun_moon.dart
import 'package:flutter/material.dart';

import '../state/state_container.dart';
import '../utils/constant.dart';


class SunMoon extends StatelessWidget {
  final bool? isSun;

  const SunMoon({
    Key? key,
    this.isSun,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final timeState = TimeModelBinding.of(context);
    return SizedBox(
      width: SUN_MOON_WIDTH,
      child: AnimatedSwitcher(
        switchInCurve: Curves.ease,
        switchOutCurve: Curves.ease,
        duration: const Duration(milliseconds: 250),
        child: isSun!
            ? Container(
                key: const ValueKey(1),
                child: timeState.widget.sunAsset ??
                    const Image(
                      image: AssetImage(
                        "packages/interval_time_selector/assets/icons/sun.png",
                      ),
                    ))
            : Container(
                key: const ValueKey(2),
                child: timeState.widget.moonAsset ??
                    const Image(
                      image: AssetImage("packages/interval_time_selector/assets/icons/moon.png"),
                    ),
              ),
        transitionBuilder: (child, anim) {
          return ScaleTransition(
            scale: anim,
            child: FadeTransition(
              opacity: anim,
              child: SlideTransition(
                position: anim.drive(
                  Tween(
                    begin: const Offset(0, 4),
                    end: const Offset(0, 0),
                  ),
                ),
                child: child,
              ),
            ),
          );
        },
      ),
    );
  }
}


lib/show_time_picker/state/state_container.dart
// ignore_for_file: must_be_immutable, no_leading_underscores_for_local_identifiers

import 'package:flutter/material.dart';

import '../utils/constant.dart';
import 'time.dart';

class _ModelBindingScope extends InheritedWidget {
  final TimeModelBindingState modelBindingState;

  const _ModelBindingScope({
    Key? key,
    required this.modelBindingState,
    required Widget child,
  }) : super(key: key, child: child);

  @override
  bool updateShouldNotify(_ModelBindingScope oldWidget) => true;
}

/// Stateful [Widget] for [InheritedWidget]
class TimeModelBinding extends StatefulWidget {
  Time selectedTime;

  final void Function(TimeOfDay) onChange;

  final void Function(DateTime)? onChangeDateTime;

  final void Function()? onCancel;

  final bool is24HrFormat;

  final bool? displayHeader;

  final Color? accentColor;

  final Color? unselectedColor;

  String cancelText;

  String okText;

  final Image? sunAsset;

  final Image? moonAsset;

  final bool blurredBackground;

  final double? borderRadius;

  final double? elevation;

  final EdgeInsets? dialogInsetPadding;

  final MinuteInterval? minuteInterval;

  bool? disableMinute;
  double minMinuteAtMinimumHour;

  final bool? disableHour;

  final double? maxHour;
  double maxMinute;

  final double? minHour;

  double minMinute;

  final String? hourLabel;

  final String? minuteLabel;

  final bool isInlineWidget;

  final bool isOnValueChangeMode;

  final bool focusMinutePicker;

  final bool ltrMode;

  TextStyle okStyle;

  TextStyle cancelStyle;

  ButtonStyle? buttonStyle;

  ButtonStyle? cancelButtonStyle;

  double? buttonsSpacing;
  double maxMinuteAtMaximumHour = 0;
  List<int> workingHours;
  final Widget child;

  TimeModelBinding({
    Key? key,
    required this.workingHours,
    required this.child,
    required this.selectedTime,
    required this.onChange,
    this.onChangeDateTime,
    this.onCancel,
    this.is24HrFormat = false,
    this.displayHeader,
    this.accentColor,
    this.ltrMode = true,
    this.unselectedColor,
    this.cancelText = "cancel",
    this.okText = "ok",
    this.isOnValueChangeMode = false,
    this.sunAsset,
    this.moonAsset,
    this.blurredBackground = false,
    this.borderRadius,
    this.elevation,
    this.dialogInsetPadding,
    this.minuteInterval,
    this.disableMinute,
    this.disableHour,
    this.maxHour,
    this.maxMinute = 55,
    this.minMinute = 0,
    this.minHour,
    required this.maxMinuteAtMaximumHour,
    required this.minMinuteAtMinimumHour,
    this.hourLabel,
    this.minuteLabel,
    this.isInlineWidget = false,
    this.focusMinutePicker = false,
    this.okStyle = const TextStyle(fontWeight: FontWeight.bold),
    this.cancelStyle = const TextStyle(fontWeight: FontWeight.bold),
    this.buttonStyle,
    this.cancelButtonStyle,
    this.buttonsSpacing,
  }) : super(key: key);

  @override
  TimeModelBindingState createState() => TimeModelBindingState();

  static TimeModelBindingState of(BuildContext context) {
    final _ModelBindingScope scope =
        context.dependOnInheritedWidgetOfExactType<_ModelBindingScope>()!;
    return scope.modelBindingState;
  }
}

class TimeModelBindingState extends State<TimeModelBinding> {
  late Time time = widget.selectedTime;
  late double hourIndex = 0;
  late double minuteIndex = 0;

  bool hourIsSelected = true;

  DayPeriod lastPeriod = DayPeriod.am;

  //
  late final double constMaxMinute;
  late final double constMinMinute;

  @override
  void initState() {
    bool _hourIsSelected = true;
    constMaxMinute = widget.maxMinute;
    constMinMinute = widget.minMinute;
    if (widget.focusMinutePicker || widget.disableHour!) {
      _hourIsSelected = false;
    }

    setState(() {
      hourIsSelected = _hourIsSelected;
    });
    super.initState();
  }

  bool didPeriodChange() {
    return lastPeriod != time.period;
  }

  void onAmPmChange(DayPeriod e) {
    setState(() {
      lastPeriod = time.period;
      time = time.setPeriod(e);
    });
  }

  onTimeChange(double value) {
    if (hourIsSelected) {
      onHourChange(value);
    } else {
      onMinuteChange(value);
    }
  }

  void onHourChange(double value) {
    if (value <= widget.workingHours.length) {
      time = time.replacing(hour: (widget.workingHours[value.toInt()]).round());
      setState(() {});
    }
  }

  void onMinuteChange(double value) {
    time = time.replacing(minute: value.ceil());
    setState(() {});
  }

  void changeMinMinute(double minMinute) {
    widget.minMinute = minMinute;
  }

  void changeMaxMinute({required double maxMinute}) {
    widget.maxMinute = maxMinute;
  }

  /// Change handler for `hourIsSelected`
  void onHourIsSelectedChange(bool newValue) {
    setState(() {
      hourIsSelected = newValue;
    });
  }

  /// [onChange] handler. Return [TimeOfDay]
  onOk() {
    widget.onChange(time.toTimeOfDay());
    if (widget.onChangeDateTime != null) {
      final now = DateTime.now();
      final dateTime =
          DateTime(now.year, now.month, now.day, time.hour, time.minute);
      widget.onChangeDateTime!(dateTime);
    }
    onCancel(result: time.toTimeOfDay());
  }

  /// Handler to close the picker
  onCancel({var result}) {
    if (widget.onCancel != null) {
      widget.onCancel!();
      return;
    }

    if (!widget.isInlineWidget) {
      Navigator.of(context).pop(result);
    }
  }

  /// Check if time is within range.
  /// Used to disable `AM/PM`.
  /// Example: if user provided [minHour] as `9` and [maxHour] as `21`,
  /// then the user should only be able to toggle `AM/PM` for `9am` and `9pm`
  bool checkIfWithinRange(DayPeriod other) {
    final tempTime = Time(time.hour, time.minute).setPeriod(other);
    final expectedHour = tempTime.hour;
    return widget.minHour! <= expectedHour && expectedHour <= widget.maxHour!;
  }

  @override
  Widget build(BuildContext context) {
    return _ModelBindingScope(
      modelBindingState: this,
      child: widget.child,
    );
  }
}


lib/show_time_picker/widgets/android_builder/day_time_picker_android.dart
import 'dart:ffi';

import 'package:flutter/material.dart';
import 'package:interval_time_selector/show_time_picker/state/time.dart';
import '../../state/state_container.dart';
import '../../utils/utils.dart';
import '../am_pm.dart';
import '../../common/action_button.dart';
import '../../common/display_value.dart';
import '../../common/filter_wraper.dart';
import '../../common/wraper_container.dart';
import '../../common/wraper_dialog.dart';
import '../day_night_banner.dart';

class DayNightTimePickerAndroid extends StatefulWidget {
  const DayNightTimePickerAndroid({Key? key}) : super(key: key);

  @override
  DayNightTimePickerAndroidState createState() =>
      DayNightTimePickerAndroidState();
}

class DayNightTimePickerAndroidState extends State<DayNightTimePickerAndroid> {
  @override
  Widget build(BuildContext context) {
    final timeState = TimeModelBinding.of(context);

    double min = 0;

    double max =
        getMaxMinute(timeState.widget.maxMinute, timeState.widget.minMinute);
    int divisions = max.toInt();

    /// if current select is hour
    if (timeState.hourIsSelected) {
      min = 0;
      max = timeState.widget.workingHours.length.toDouble() - 1;
      divisions = max.toInt();
      if (timeState.widget.workingHours.contains(timeState.time.hour)) {
        timeState.hourIndex = timeState.widget.workingHours
            .indexOf(timeState.time.hour)
            .toDouble();
      } else {
        timeState.hourIndex = 0;
        timeState.time = Time(
            timeState.widget.workingHours[timeState.hourIndex.toInt()],
            timeState.widget.minMinuteAtMinimumHour.toInt());
      }

      timeState.hourIndex = timeState.hourIndex == -1 ? 0 : timeState.hourIndex;
    }

    /// if current select is minute
    else {
      if (timeState.hourIndex == 0) {
        timeState.widget.minMinute = timeState.widget.minMinuteAtMinimumHour;
        max = getMaxMinute(
            timeState.widget.maxMinute, timeState.widget.minMinute);
      }
      if (timeState.widget.workingHours[timeState.hourIndex.toInt()] ==
          timeState.widget.maxHour) {
        max = getMaxMinute(timeState.widget.maxMinuteAtMaximumHour,
            timeState.widget.minMinute);
      }
      divisions = max.toInt();
      timeState.minuteIndex =
          ((timeState.time.minute - timeState.widget.minMinute) /
              getIntFromMinuteIntervalEnum(timeState.widget.minuteInterval));
    }

    final color =
        timeState.widget.accentColor ?? Theme.of(context).colorScheme.secondary;

    final hourValue = timeState.widget.is24HrFormat
        ? timeState.time.hour
        : timeState.time.hourOfPeriod;

    final ltrMode =
        timeState.widget.ltrMode ? TextDirection.ltr : TextDirection.rtl;

    Orientation currentOrientation = MediaQuery.of(context).orientation;
    return Center(
      child: SingleChildScrollView(
        physics: currentOrientation == Orientation.portrait
            ? const NeverScrollableScrollPhysics()
            : const AlwaysScrollableScrollPhysics(),
        child: FilterWrapper(
          child: WrapperDialog(
            child: Column(
              mainAxisSize: MainAxisSize.min,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: <Widget>[
                WrapperContainer(
                  child: Column(
                    mainAxisSize: MainAxisSize.min,
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: <Widget>[
                      const AmPm(),
                      Expanded(
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: <Widget>[
                            DisplayValue(
                              onTap: timeState.widget.disableHour!
                                  ? null
                                  : () {
                                      timeState.onHourIsSelectedChange(true);
                                    },
                              value: hourValue.toString().padLeft(2, '0'),
                              isSelected: timeState.hourIsSelected,
                            ),
                            const DisplayValue(
                              value: ":",
                            ),
                            DisplayValue(
                              onTap: timeState.widget.disableMinute!
                                  ? null
                                  : () {
                                      if (max == min) {
                                        timeState.changeMaxMinute(
                                            maxMinute:
                                                timeState.widget.maxMinute);
                                      }
                                      timeState.onHourIsSelectedChange(false);
                                    },
                              value: timeState.time.minute
                                  .toString()
                                  .padLeft(2, '0'),
                              isSelected: !timeState.hourIsSelected,
                            ),
                          ],
                        ),
                      ),
                      min != max
                          ? Slider(
                              onChangeEnd: (value) {
                                if (timeState.widget.isOnValueChangeMode) {
                                  timeState.onOk();
                                }
                              },
                              value: timeState.hourIsSelected
                                  ? timeState.hourIndex
                                  : timeState.minuteIndex,
                              onChanged: (value) {
                                value = value.ceil().toDouble();
                                if (value == timeState.hourIndex &&
                                    timeState.hourIsSelected) {
                                  return;
                                } else if (value == timeState.minuteIndex &&
                                    !timeState.hourIsSelected) {
                                  return;
                                }

                                if (timeState.hourIsSelected) {
                                  timeState.hourIndex = value;
                                } else {
                                  timeState.minuteIndex = value;
                                  value = (value *
                                          getIntFromMinuteIntervalEnum(timeState
                                              .widget.minuteInterval)) +
                                      timeState.widget.minMinute;
                                }

                                timeState.onTimeChange(value);

                                if (timeState.hourIsSelected) {
                                  /// selected hour equal to maximum hour check add allowed maximum minute to max minut
                                  if (value == max) {
                                    timeState.changeMinMinute(
                                        timeState.constMinMinute);
                                    timeState.onMinuteChange(
                                        timeState.constMinMinute);
                                    timeState.minuteIndex =
                                        timeState.constMinMinute;
                                    timeState.changeMaxMinute(
                                        maxMinute: timeState
                                            .widget.maxMinuteAtMaximumHour);
                                  }

                                  /// selected hour is current hour change min minute
                                  else if (value == min) {
                                    timeState.changeMinMinute(timeState
                                        .widget.minMinuteAtMinimumHour);
                                    timeState.onMinuteChange(timeState
                                        .widget.minMinuteAtMinimumHour);

                                    timeState.changeMaxMinute(
                                        maxMinute: timeState.constMaxMinute);
                                  } else {
                                    timeState.changeMinMinute(
                                        timeState.constMinMinute);
                                    timeState.onMinuteChange(
                                        timeState.constMinMinute);
                                    timeState.minuteIndex =
                                        timeState.constMinMinute;
                                    timeState.changeMaxMinute(
                                        maxMinute: timeState.constMaxMinute);
                                  }
                                }
                              },
                              min: min,
                              max: max,
                              divisions: divisions > 0 ? divisions : null,
                              activeColor: color,
                              inactiveColor: color.withAlpha(55),
                            )
                          : const SizedBox(),
                      const ActionButtons(),
                    ],
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:interval_time_selector/show_time_picker/day_time_picker.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Time picker',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  TimeOfDay _time = TimeOfDay.now().replacing(hour: 14, minute: 30);
  bool iosStyle = true;

  void onTimeChanged(TimeOfDay newTime) {
    setState(() {
      _time = newTime;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Center(
          child: SingleChildScrollView(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Text(
                  "Popup Picker Style",
                  style: Theme.of(context).textTheme.headline6,
                ),
                Text(
                  _time.format(context),
                  textAlign: TextAlign.center,
                  style: Theme.of(context).textTheme.headline1,
                ),
                const SizedBox(height: 10),
                TextButton(
                  style: TextButton.styleFrom(
                    backgroundColor: Theme.of(context).colorScheme.secondary,
                  ),
                  onPressed: () {
                    Navigator.of(context).push(
                      showPicker(
                        workingHours: [0, 1, 9, 12, 11, 12, 13, 15, 19, 23],
                        context: context,
                        value: _time,
                        ascending: true,
                        onChange: onTimeChanged,
                        maxMinuteAtMaximumHour: 40,
                        minMinuteAtMinimumHour: 20,
                        onChangeDateTime: (DateTime dateTime) {
                          debugPrint("[debug datetime]:  $dateTime");
                        },
                      ),
                    );
                  },
                  child: const Text(
                    "Open time picker",
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
